[1단계 - 로또 구현] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 앤지! 👋
이번 1단계 미션 리뷰를 받게된 도기입니다! 리뷰에 참여해주셔서 감사합니다🙇‍♂️
[셀프 체크](https://prolog.techcourse.co.kr/studylogs/3877)

이번 미션을 진행하면서 다음과 같은 고민들을 하게 되었습니다.

# 고민 포인트 

## 1. static 키워드에 대한 기준은?
페어와 미션 진행중 메서드에 static 키워드를 붙이는 기준에 대해서 의견차이가 있었습니다. static 키워드를 사용하는 저의 기준은 아래와 같습니다.
### 많은 곳에서 참조가 필요할 경우 static 키워드를 사용해도 좋다.
- static 키워드를 사용하면 인스턴스를 생성하지 않고 사용할 수 있기 때문에 속도가 빠르다는 장점이 있습니다.
- 프로그램 실행시 메모리에 적재되기 때문에 다양한 곳에서 빈번하게 참조되는 메서드라면 메모리를 효율적으로 사용할 수 있습니다.

하지만 static 키워드는 객체의 상태에 영향을 받지 않기 때문에 `객체지향적이지 않다` 라는 단점이 있습니다. 또한 무분별하게 static을 사용하게 되면 메모리를 효율적으로 사용하지 못할 것 이라는 생각이 듭니다.

이런 부분에 있어서 앤지의 생각은 어떠신지 궁금합니다! 
또한 static 키워드를 사용하는 앤지만의 기준은 무엇인가요? 왜 static 키워드를 사용하시나요? 제가 알고있는 static에 대한 지식이 잘못되었다면 지적해주시면 감사하겠습니다!

## 2. 메서드 분리의 기준
미션의 프로그래밍 요구사항중 `인덴트를 1로 제한한다`라는 요구사항이 있었습니다. 
요구사항을 지키기위해서 페어와 의견이 갈리는 부분이 있었습니다.

``` Java
@Override
public List<Integer> pickUnique(int start, int end, int count) {
    List<Integer> result = new ArrayList<>();

    while (result.size() < count) {
        int random = this.random.nextInt(end) + start;
        if (!result.contains(random)) {
            result.add(random);
        }
    }

    return Collections.unmodifiableList(result);
}
```
위 메서드에서 제가 생각하는 메서드 분리의 기준은
- 메서드의 변경 가능성
- 메서드가 한가지 일만 하고 있는가?

입니다.

저의 기준에 따라서 메서드가 하는 일을 분리해본다면

- count 만큼 반복한다.
- 랜덤한 번호를 생성한다.
- 이미 중복된 숫자가 있는지 검사한다.
- 랜덤한 숫자를 리스트에 삽입한다.

로 구분할 수 있을 것 같습니다.

반대로 페어의 의견은 
- `인덴트가 2단계까지 사용되었지만, '중복없는 숫자를 생성한다'라는 하나의 동작을 충분히 잘 수행하고 있다` 

즉 페어는 `비즈니스로직을 수행하기 위한 동작 하나하나가 중요하고, 편하게 한국어로 읽히는것이 중요하다고 생각하기 때문에 메서드 분리는 과한 선택이다` 라는 의견이었습니다.

이러한 메서드 분리 기준에 대한 앤지의 생각은 어떤가요? 앤지만의 메서드 분리 기준은 어떻게 되나요? 어디까지가 분리의 기준이라고 생각하시나요?

# 아쉬운 부분
## 도메인 객체 분리의 아쉬움
초반 설계의 공을 들이지 않고 바로 구현에 돌입하여 구조에 신경쓰지 못했습니다. 그러한 이유 때문에 빈번하게 수정이 일어났고 결론적으로 `Lottos` 객체에 너무 많은 책임이 부여된것 같은 느낌이 들었습니다.

## 작성한 기능 목록을 기간내에 구현하지 못한 것
잔돈을 거슬러주는 기능을 기능 목록사항에 추가했지만 마감 시간이 임박해 구현하지 못하였습니다. 임의로 정한 요구사항이었지만 시간 분배를 잘 못했다는 것에 아쉬움이 남는 것 같습니다.

이번 리뷰에 참여해주셔서 감사드립니다! 🙇‍♂️

dompoo committed:
docs: 기능 요구사항 추가

dompoo committed:
feat: 로또 구입 금액 입력 기능 개발

dompoo committed:
feat: 랜덤한 번호 생성 기능 개발

dompoo committed:
feat: 로또 구매 기능 개발

dompoo committed:
feat: 구입한 로또 개수 출력 기능 개발

dompoo committed:
test: 중복 없는 번호 생성 기능 테스트 추가

dompoo committed:
test: 구입 금액만큼 로또 구매 테스트 추가

dompoo committed:
feat: 생성된 로또 번호 출력 기능 개발

dompoo committed:
feat: 숫자 뽑기 기능을 NumberPicker로 추상화

- 테스트 상황에서 로또 번호를 테스트하기 위해, NumberPicker 인터페이스로 추상화
- StaticNumberPicker를 통해 테스트에서 정해진 값으로 테스트 가능

dompoo committed:
test: 구입 금액만큼 로또 구입 테스트에서 생성된 로또 번호도 테스트

dompoo committed:
feat: 당첨 로또번호 입력 기능 개발

dompoo committed:
feat: 보너스 번호 입력 기능 개발

dompoo committed:
feat: 로또 당첨 등수를 계산하는 기능 개발

dompoo committed:
feat: 당첨번호개수와 보너스번호 당첨 여부를 계산하는 기능 개발

dompoo committed:
feat: 로또 당첨 통계를 계산하는 기능 개발

dompoo committed:
fix: LottoPrize에서 필요 당첨 개수와 보너스볼 일치 여부를 확인할 수 없는 문제 해결

- 람다식 대신에 필요 당첨 개수와 보너스볼 일치 여부를 직접 갖고 있게 수정

dompoo committed:
fix: 입력시 개행문자 포함하도록 수정

- 정수입력 기대상황에서 개행문자를 무시하여 정수입력이 아닌 문자로 입력되는 문제를 해결

dompoo committed:
refactor: 로또등수 상수 순서 변경

dompoo committed:
feat: 당첨 통계 출력 기능 개발

dompoo committed:
feat: 수익률 계산/출력 기능 개발

- 기존 LottoStaticCalculaator를 Lottos로 이름바꾸고, 역할을 Lotto에 관한 통계/수익률 계산으로 두었습니다.
- 이를 통해 Lotto를 사용하기 더 편해졌습니다.

dompoo committed:
feat: 금액을 입력하지 않은 경우 예외처리 추가

dompoo committed:
feat: 금액을 숫자로 입력하지 않은 경우 예외처리 추가

dompoo committed:
feat: 금액을 1000원 미만으로 입력한 경우 예외처리 추가

dompoo committed:
feat: 로또 생성기능 예외처리 추가

dompoo committed:
feat: 로또 당첨 번호를 쉼표로 구분하지 않은 경우 예외처리 추가

dompoo committed:
feat: 로또 당첨 번호에 중복이 있는 경우 예외처리 추가

dompoo committed:
refactor: 로또에서 당첨 로또 번호, 보너스 번호와 비교한 결과를 한번에 반환하도록 리팩터링

- 기존 코드에서는 각각 반환하기 때문에 보너스 번호가 당첨 로또 번호와 중복되어도 탐지할 수 없었습니다.
- 이제 한 메서드 내에서 두 값을 비교하여 중복을 감지할 수 있습니다.

dompoo committed:
feat: 보너스 번호가 당첨 로또 번호와 중복된 경우 예외처리 추가

dompoo committed:
feat: 로또 당첨 번호가 6개가 아닌 경우 예외처리 추가

dompoo committed:
refactor: 15라인이 넘는 메서드 분리 리팩터링

- 가독성이 좋지않은 메서드를 연관성이 있는 메서드끼리 분리했습니다.

dompoo committed:
refactor: 뜻이 불명확한 값을 상수화

dompoo committed:
refactor: 인덴트가 많아 불명확한 코드 리팩터링

dompoo committed:
chore: editorconfig 추가

soominsohn commented on src/main/java/controller/InputView.java:
```diff
@@ -0,0 +1,63 @@
+    
+    public static final char MATCH_LOTTO_DELIMITER = ',';
+    
+    public final Scanner SC;
```
Scanner를 상수로 선언하시려고 한걸까요? 아니면 필드로 생성하려고 하신걸까요?
보통 멤버 필드는 카멜케이스로 선언합니다!

dye0p replied:
엇..! 이 부분은 필드로 생성하려고 했는데 실수한것 같습니다..ㅠ

soominsohn commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,104 @@
+    private static final int NUMBER_COUNT = 6;
+    private static final int PRICE = 1000;
+    
+    private final List<Integer> numbers;
```
로또 번호도 도메인으로 관리해보면 어떨까요?

dye0p replied:
로또 번호에 대한 일급 컬렉션을 생성하면 좋겠다는 말씀이신가요?~ 반영해보겠습니다!😊

soominsohn commented on src/main/java/domain/LottoPrize.java:
```diff
@@ -0,0 +1,67 @@
+ */
+public enum LottoPrize {
+    
+    FIFTH(3, null, 5_000),
+    FOURTH(4, null, 50_000),
+    THIRD(5, false, 1_500_000),
+    SECOND(5, true, 30_000_000),
+    FIRST(6, null, 2_000_000_000),
+    ;
+    
+    private final int minMatchCount;
+    private final Boolean isBonusMatch;
+    private final long prizeMoney;
+    
```
isBonusMatch 필드에 Boolean Wrapper Type을 쓰셨군요! 
이유가 따로 있을까요?
primitive type vs wrapper type에 대해 학습해보고 적절히 수정해보면 좋을 것 같아요 😄 

[참고](https://velog.io/@kimdy0915/%EA%B8%B0%EB%B3%B8%ED%98%95primitive-vs.-%EB%9E%98%ED%8D%BC-%ED%81%B4%EB%9E%98%EC%8A%A4wrapper-class)

dye0p replied:
isBonusMatch 필드에 래퍼타입을 적용한 이유는 보너스볼 당첨 유무가 중요한 2등과 3등을 제외한 나머지 등수에 대해서는 null을 저장해서 보너스볼 당첨 유무의 중요성을 없애기 위함이었습니다!
올려주신 아티클을 참고해서 기본형(boolean) 으로 구현할 수 있을지 고민해본 후 반영해보겠습니다!

soominsohn commented on src/main/java/util/RandomNumberPicker.java:
```diff
@@ -0,0 +1,69 @@
+        if (!result.contains(random)) {
+            result.add(random);
+        }
+    }
```
`2. 메서드 분리의 기준` 질문에 대한 코드인 것 같네요!
위 함수는 '중복 없는 숫자를 생성한다.'이라는 하나의 동작을 수행하고 반환하지만, 메서드 내부적으로는 여러 작업을 하고 있네요. 
- 랜덤 숫자 생성
- 중복 검사
- 리스트 추가
이 모든 동작이 한 메서드안에 들어있다면 처음 보는 사람이 파악하기 어려울 수 있을 것 같아요.

예를들어 랜덤 숫자를 생성하는 `this.random.nextInt(end) + start;` 이 코드만 본다면 '랜덤 숫자를 생성한다' 라는 기능을 한눈에 파악하기 어렵지만 'creatRandomNumber'라고 메서드를 분리해주면 명확히 파악할 수 있을 것 같습니다.
좋은 토론을 하신 것 같아요! 😄

soominsohn commented on README.md:
```diff
@@ -1,3 +1,50 @@
 
 로또 미션 저장소
+
+## 기능 요구사항
```
요구사항 잘 정리해주셨네요 😄

soominsohn commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,139 @@
+
+class LottoTest {
+    
+    @Test
+    void 로또_번호가_6개가_아니면_예외가_발생한다() {
+        List<Integer> numbers = List.of(1, 2, 3, 4, 5);
+        
+        assertThatThrownBy(() -> new Lotto(numbers))
+                .isExactlyInstanceOf(IllegalArgumentException.class)
+                .hasMessage("로또 번호는 6개가 되어야 합니다.");
+    }
```
테스트 작성 👍 
지금은 예외 케이스들에 대해서만 작성되어있는데, 정상케이스에 대해서도 잘 동작하는지 작성해보면 좋을 것 같아요.

dye0p replied:
넵 반영해보겠습니다!

soominsohn commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,104 @@
+        validateBonusNumberNotDuplicated(matchNumbers, bonusNumber);
+        int matchCount = 0;
+        
+        for (Integer matchNumber : matchNumbers) {
+            if (numbers.contains(matchNumber)) {
+                matchCount++;
+            }
+        }
```
`규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.`를 적용해볼까요?

dye0p replied:
앗 검토가 부족했네요ㅠㅠ 반영해보겠습니다👍

soominsohn commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+    
+    public static List<Lotto> purchase(int money, NumberPicker numberPicker) {
+        validateMinPrice(money);
+        
+        List<Lotto> lottos = new ArrayList<>();
+        int lottoCount = money / PRICE;
+        for (int i = 0; i < lottoCount; i++) {
+            List<Integer> numbers = numberPicker.pickUnique(START_NUMBER, END_NUMBER, NUMBER_COUNT);
+            Lotto lotto = new Lotto(numbers);
+            lottos.add(lotto);
+        }
+        return Collections.unmodifiableList(lottos);
+    }
```
Lotto 객체 안에서 Lotto 리스트를 생성해서 반환하네요!
도기는 처음 Lotto 객체를 설계하실 때 어떤 역할을 하는 객체라고 생각을 하셨나요?
Lotto 리스트를 생성해서 반환하는게 Lotto의 책임일지 생각해보면 좋겠습니다 😄

dye0p replied:
처음 구현을 할때는 로또에게 구매금액을 주고 자연스럽게 로또들이 생성되면 좋겠다고 생각했습니다! 조금 더 깊게 생각해보니 로또들을 생성하는 객체를 생성하면 좋을것 같다는 생각이드는데 이 부분에 대해서 한번 수정해보겠습니다! 추가로 이 부분에 대해서 앤지의 생각은 어떤지 궁금합니다!😊

soominsohn replied:
네~ 저도 로또에서 로또들을 알 필요는 없다고 생각합니다~! `LottoMachine` 잘 만들어주셨네요 😄

soominsohn commented on src/main/java/domain/LottoPrize.java:
```diff
@@ -0,0 +1,67 @@
+        this.prizeMoney = prizeMoney;
+    }
+    
+    public static Optional<LottoPrize> match(LottoMatchResult matchResult) {
```
dto 패키지 하위에있는 LottoMatchResult를 domain 에서 사용하고 계시군요!
LottoMatchResult를 dto로 만들어주신 이유가 따로 있을까요?

dye0p replied:
기존 코드에서는 당첨 번호와 보너스 당첨 여부를 각각 반환하였기 때문에 보너스 번호가 당첨 로또 번호와 중복이 되어도 탐지할 수 없었습니다. 그래서 당첨 번호 일치 개수와 보너스번호 당첨 여부를 한번에 반환하기 위해서 dto를 사용해서 필요한 값을 한번에 받을 수 있도록 하였습니다! dto 사용에 대한 앤지의 생각도 궁금해요!🤔

soominsohn replied:
네~ 저도 당첨 번호 일치 개수와 보너스 당첨 여부를 필드로 가지고 있는 LottoMatchResult 객체가 있어야 한다고는 생각합니다.
다만 LottoMatchResult가 dto 패키지 하위에 있는걸로 보아 도메인이 아닌 dto라고 생각하신 것 같아서 물어본 겁니다.
dto를 `LottoPrize`, `Lotto`, `Lottos`와 같은 도메인 객체에서 사용하고 계신데 
- 과연 도메인에서 dto의 존재를 알고 있어야할까요?
- LottoMatchResult는 도메인일까요? dto일까요?  

dto의 역할이 무엇인지 학습해보고 적절히 수정해보면 좋을 것 같습니다. 궁금한 점 있으면 2단계 하시면서 질문주세요~!

soominsohn commented on src/main/java/util/RandomNumberPicker.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.List;
+import java.util.Random;
+
+public class RandomNumberPicker implements NumberPicker {
+    
+    private final Random random;
+    
+    public RandomNumberPicker(Random random) {
+        this.random = random;
+    }
+    
```
Number picker 인터페이스를 정의하고, 인터페이스를 구현하는 RandomNumberPicker를 사용하셨네요! 
어떤 이유로 만드셨고 어떤 장점이 있었을까요?

dye0p replied:
랜덤값을 테스트하기 위함이었어요! 랜덤한 로또번호를 임의의 로또 번호로 테스트하기 위해서는 인터페이스를 정의해서 입력한 번호를 그대로 반환하는 테스트용 구현체를 만들면 임의의 값으로 테스트 할 수 있겠다고 생각했어요! 이렇게 하게되면 정확한 테스트가 가능하다는 장점이 있다고 생각합니다ㅎㅎ 앤지는 어떻게 생각하시나요?

soominsohn replied:
네~ 이 방식은 `전략 패턴` 활용한 좋은 방식이에요!
도기가 말씀해주신 내용과 더불어 로또 생성 정책이 추가되거나 바뀔 때 장점이 하나 더 생기는데요 2단계 해보시면 몸소 체험하실 수 있겠네요 ㅎㅎ

soominsohn commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,104 @@
+            throw new IllegalArgumentException("로또 번호는 중복되면 안됩니다.");
+        }
+    }
+    
+    public static List<Lotto> purchase(int money, NumberPicker numberPicker) {
+        validateMinPrice(money);
+        
```
1번 질문에 대한 답변을 여기 달면 적절할 것 같군요. 😄 
전 주로 static 메서드는 유틸성 메서드처럼 여러 곳에서 공통적으로 사용되는 기능을 담고 있거나, 객체 상태와 관계없는 기능을 하는 메서드에 사용합니다. 
도기가 static 메서드에 대해 잘 정리해주셨는데 앞으로 더 많은 코드를 작성하다보면 감이 오실거에요. 
그러나 Lotto 객체 안에 로또 리스트를 만들기 위한 static 메서드는 적절해보이지 않습니다. 한번 생각해보고 개선해보면 좋을 것 같아요 😄

dye0p replied:
좋은 의견 감사합니다ㅎㅎ 말씀해주신 의견을 참고해서 수정해보겠습니다!

soominsohn reviewed (changes requested):
도기 안녕하세요!
이번 리뷰를 맡게된 앤지라고합니다. 잘 부탁드려요!
첫 미션인데 정말 잘해주셨습니다. 
주신 질문에 대한 답과 코멘트 몇개 남겼으니 확인 부탁드립니다 😄

dye0p committed:
refactor: Scanner 필드명 소문자로 변경

dye0p committed:
refactor: 로또 번호를 가지는 일급 컬렉션Numbers 객체 추가

- 기존 List로 관리되면 로또 번호를 일급 컬렉션인 Numbers 객체로 포장
- 기존의 테스트 코드 수정
- 로또 번호와 관련된 책임을 Numbers 객체로 이동
- Numbers 테스트 완료

dye0p committed:
test: Numbers 객체 정상 테스트 추가

- Numbers 생성 정상 테스트 추가
- 디스플레이 네이밍으로 테스트명 변경
- given,when,than 절 추가

dye0p committed:
refactor: numbers 객체의 contains 메서드 네이밍 직관적으로 변경

dye0p committed:
refactor: 당첨 번호 카운팅 로직 메서드 분리 리팩터링

- 당첨 번호 카운팅 로직을 분리하여 메서드로 행동을 명확하게 표현 하였음.

dye0p committed:
feat: 로또를 발행하는 LottoMachine 객체를 추가

- 기존 Lotto 객체에서 로또를 발행하는 것을 로또를 발행하는 LottoMachine 객체를 생성하여 로또 발행 책임 이동

- 로또 발행 테스트 이동
- 실패, 성공 테스트 추가

dye0p committed:
refactor: LottoPrize enum 객체의 isBonusMatch 필드 타입을 기본형으로 변경

- 래퍼타입으로 필드를 사용했을때의 메모리 누수 문제를 방지 하기 위해 기본형 타입으로 변경하였음
- null 발생에 대한 안전성 보장

dye0p committed:
refactor: 로또 발행시 Lottos로 반환되도록 변경

dye0p committed:
fix: 구매한 로또 번호의 출력이 안되던 문제 해결

- 구매한 로또 번호 출력시 객체의 주소가 출력되었던 문제를 toString() 메서드를 사용하여 숫자로 출력하도록 수정

dye0p commented:
안녕하세요 앤지! 피드백  주신 것을 바탕으로 수정하여 다시 리뷰 요청합니다! 수고 하셨습니다 :)

soominsohn reviewed (approved):
안녕하세요 도기~!
1단계 정말 잘 해주셨어요~
코멘트에 대한 답변 남겨두었으니 확인해주시고 2단계도 파이팅입니다~
